pkg/gencert/gencert.go

```diff
@@ -0,0 +1,179 @@
+limitations under the License.
+*/
+
+package gencert
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	certificates "k8s.io/api/certificates/v1beta1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/keyutil"
+	"time"
+)
+
+const (
+	// systemPrivilegedGroup is a superuser by default (i.e. bound to the cluster-admin ClusterRole).
+	systemPrivilegedGroup = "system:masters"
+	// certWaitInterval certificate request poll interval.
+	certWaitInterval = 1 * time.Second
+	// certWaitTimeout certificate request poll timeout.
+	certWaitTimeout = 20 * time.Second
+)
+
+// generateCSR generates a certificate signing request (CSR).
+func generateCSR() (*certificates.CertificateSigningRequest, []byte, error) {
+
+	// Generate a new private key.
+	pk, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to generate a new private key: %v", err)
+	}
+
+	// Marshal pk -> der.
+	der, err := x509.MarshalECPrivateKey(pk)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to marshal the new key to DER: %v", err)
+	}
+
+	// Generate PEM key.
+	keyPEM := pem.EncodeToMemory(&pem.Block{Type: keyutil.ECPrivateKeyBlockType, Bytes: der})
+
+	// Generate a x509 certificate signing request.
+	csrPEM, err := cert.MakeCSR(pk, &pkix.Name{CommonName: "client", Organization: []string{systemPrivilegedGroup}}, nil, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to create a csr from the private key: %v", err)
+	}
+
+	// Generate a Kubernetes CSR object.
+	csrObj := &certificates.CertificateSigningRequest{
+		// Username, UID, Groups will be injected by API server.
+		TypeMeta: metav1.TypeMeta{Kind: "CertificateSigningRequest"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:         "",
+			GenerateName: "csr-",
+		},
+		Spec: certificates.CertificateSigningRequestSpec{
+			Request: csrPEM,
+			Usages: []certificates.KeyUsage{
+				certificates.UsageDigitalSignature,
+				certificates.UsageKeyEncipherment,
+				certificates.UsageClientAuth,
+			},
+		},
+	}
+
+	return csrObj, keyPEM, nil
+}
+
+// requestCSR requests a certificate signing request (CSR).
+func requestCSR(clientset kubernetes.Interface, csrObj *certificates.CertificateSigningRequest) ([]byte, error) {
+	client := clientset.CertificatesV1beta1().CertificateSigningRequests()
+
+	// Create CSR.
+	csrObj, err := client.Create(csrObj)
+	if err != nil {
+		return nil, fmt.Errorf(" failed to create CSR: %v", err)
+	}
+
+	csrName := csrObj.Name
+
+	// Approve CSR.
+	err = wait.Poll(certWaitInterval, certWaitTimeout, func() (bool, error) {
+		appendApprovalCondition(csrObj)
+		csrObj, err = client.UpdateApproval(csrObj)
+		if err != nil {
+			return false, nil
```

Comment:
what happens if you return `false, err`?

Reply:
I think ... if the poll times out then the error returned here will be what is returned and assigned to `err` at top-level `err = wait.Poll(...)` rather than generic timeout error ... but I am not certain. (will verify)

Reply:
What I said is indeed the scenario when using `wait.Poll`(https://godoc.org/k8s.io/apimachinery/pkg/util/wait#Poll), so it would make sense to actually return the error.

Reply:
Updated